### DISCUSSION TITLE
Allow to set execution_timeout default value in airflow.cfg

### DISCUSSION THREAD

**Discussion by alexInhert** [apache/airflow#18411] [General]
### Description

Currently the default value of execution_timeout in base operator is None
https://github.com/apache/airflow/blob/c686241f4ceb62d52e9bfa607822e4b7a3c76222/airflow/models/baseoperator.py#L502

This means that a task will run without limit till finished.

### Use case/motivation

The problem is that there is no way to overwrite this default for all dags. This causes problems where we find that tasks run sometimes for 1 week!!! for no reason. They are just stuck. This mostly happens with tasks that submit work to some 3rd party resource. The 3rd party had some error and terminated but the Airflow task for some reason did not.

This can be handled in cluster policy however it feels that cluster policy is more to enforce something but less about setting defaults values.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by BasPH** [comment]
You might want to check out task policies: https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html#task-policies

**Comment by alexInhert** [comment]
> You might want to check out task policies: https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html#task-policies

I'm aware. thanks. I mentioned it's downside on the description. Another huge downside is https://github.com/apache/airflow/issues/18253

**Comment by potiuk** [comment]
This would be a backwards incompatible-change, and it is completely needed and one that could be easily might be overridden by default_args of the DAG - which I believe should be more than enough - you can easly set default execution_timeout at the DAG level, which I believe is the only reasonable "common" setting. I think it is a very bad idea to set execution timeout on "organisation" or "group" or "whole airflow instance" level. That would make an unexpected behaviours for people who would not know that rule - that's violates "explicit is better than implicit" Python rule.

Even If you want to enforce it, then in your Task Policy you should not change value of the parameter but you should simply fail the Task if you see that it runs without the right parameter set. This is the same about #18253 (I am just going to comment on that) - you should FAIL the task rather than mutate it in this case if it violates the Policy. There are some cases where you would likely want to mutate the task, but I think vast majority of cases is to fail it when the Policy is not met. 

As opposed to "magic defaults" this has multiple good properties:

* it is very explicit
* it teaches people to use good practices (especially if your exceptions are informative, provide the user with explanation and guide them how to comply with the Policy.
* it allows you to make some more fine-grained check (like check that the timeout is <1h for certain kinds of operators and < 2h for other kinds - and you have full control over that at Airflow instance level

I am converting that issue into discussion - I would like to hear why you would not be able to use it this way., I see that as non-issue, more of a choice that the Airflow project made.

**Reply by alexInhert** [reply]
@potiuk 
It doesn't have to be backwards incompatible-change. The default value can remain like it is today.

yes we can use `default_args` but this still means that each of our users needs to remember to set the value of `execution_timeout` in his `default_args`  and even if we create a common `default_args`  we still need each user to import it. We don't have control over what each user does. this is the problem.
We just want to overwrite the default value of Airflow. To give some context  Airflow does have offer it for some properties like:
retries https://github.com/apache/airflow/blob/main/airflow/models/baseoperator.py#L485
So why retries can be a setting from `airflow.cfg` and `execution_timeout` can't?

The only way to have default value control is to allow to set it from airflow.cfg

By the way similar request is also for `weight_rule` see: https://github.com/apache/airflow/issues/12807

I think the change needs to be:
https://github.com/apache/airflow/blob/main/airflow/models/baseoperator.py#L502
to something like:

`execution_timeout: Optional[timedelta] = conf.gettimedelta('core', 'default_ execution_timeout', fallback=None),`

**Reply by potiuk** [reply]
Actually  - you are right. I am not sure why I have not seen that :). Looks like a doable change. Would you like to submit a PR for that? Seems like good one for first-time contribution?

You can make PR directly - no need to have a feature/issue for that.

**Reply by alexInhert** [reply]
I'm working on some other feature I'd like to contribute so have little time for this.
hoping someone will find it interesting and take it.

**Reply by alexInhert** [reply]
@potiuk shall I create it as issue ?

**Reply by potiuk** [reply]
I converted it back to an issue. But - just to let you know - if you need it and want it, the the easiest way to get it done is to implement it yourself (but this is a simple issue - I will mark it as good-first-issue and maybe someone indeed will pick it up).
